oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/base/StorageEsInstaller.java

```diff
@@ -38,68 +42,165 @@
-                    log.info(
-                        "create {} index template finished, isAcknowledged: {}", model.getName(), isAcknowledged);
-                    if (!isAcknowledged) {
-                        throw new StorageException("create " + model.getName() + " index template failure, ");
-                    }
-                }
-                indexName = TimeSeriesUtils.latestWriteIndexName(model);
+            boolean updateTemplate = !esClient.isExistsTemplate(tableName);
+            if (!updateTemplate) {
+                appendTemplateMapping(tableName, (Map<String, Object>) esClient.getTemplate(tableName)
+                                                                               .get("mappings"));
+                updateTemplate = !isTemplateMappingCompatible(tableName, mapping);
             }
-            if (!esClient.isExistsIndex(indexName)) {
-                boolean isAcknowledged = esClient.createIndex(indexName);
-                log.info("create {} index finished, isAcknowledged: {}", indexName, isAcknowledged);
+            if (updateTemplate) {
+                Map<String, Object> templateMapping = appendTemplateMapping(tableName, mapping);
+                boolean isAcknowledged = esClient.createOrUpdateTemplate(tableName, settings, templateMapping);
+                log.info("create {} index template finished, isAcknowledged: {}", tableName, isAcknowledged);
                 if (!isAcknowledged) {
-                    throw new StorageException("create " + indexName + " time series index failure, ");
+                    throw new IOException("create " + tableName + " index template failure, ");
                 }
-            }
 
+                if (esClient.isExistsIndex(indexName)) {
+                    Map<String, Object> historyMapping = (Map<String, Object>) esClient.getIndex(indexName)
+                                                                                       .get("mappings");
+                    Map<String, Object> appendMapping = extractAppendMapping(templateMapping, historyMapping);
+                    if (appendMapping.size() > 0) {
+                        esClient.updateIndexMapping(indexName, appendMapping);
+                    }
+                } else {
+                    isAcknowledged = esClient.createIndex(indexName);
+                    log.info("create {} index finished, isAcknowledged: {}", indexName, isAcknowledged);
+                    if (!isAcknowledged) {
+                        throw new StorageException("create " + indexName + " time series index failure, ");
+                    }
+                }
+            }
         } catch (IOException e) {
-            throw new StorageException(e.getMessage());
+            throw new StorageException("cannot create " + tableName + " index template", e);
+        }
+    }
+
+    private Map<String, Object> extractAppendMapping(final Map<String, Object> latestMapping,
```

Comment:
I hope you could shape/polish the comparing and extract extra fields/mappings in a better way. Such as using the new `class IndexFields`. Existing status and new fields should be appended in creating stage should be managed inside it.